### PR TITLE
Adding opentracing

### PR DESCRIPTION
**Pull request by samincheva**
## Changes proposed in this PR
- Adds open tracing library.
- Adds open tracing to the core parts of the match cycle logic.

## Why are we making these changes?
This will allow us to inspect runtime performance bottlenecks.

### CODE REVIEW

File: scheduler/src/cook/kubernetes/compute_cluster.clj [hunk 1]
```diff
@@ -344,21 +345,23 @@
...
-            (doall
-              (map (fn [task-metadata]
-                     (.submit
-                       controller-executor-service
-                       ^Callable (fn []
-                                   (launch-task! this task-metadata)
-                                   (process-task-post-launch-fn task-metadata))))
-                   task-metadata-seq))]
-        (run! deref futures))))
+      (tracing/with-span [s {:name "launch-tasks" :tags
```
Comment (scrosby) [inline scheduler/src/cook/kubernetes/compute_cluster.clj:348]: Looks fine, but I'm not sure I like the naming scheme of `launch-tasks`. That could mean several things, especially as we add more spans across Cook.

Maybe something like `k8s.launch-tasks`, but I'm open to suggestions.

Also, from https://github.com/opentracing/specification/blob/master/semantic_conventions.md#span-tags-table  should we consider using the standardized `component` field of `cook.k8s`? 

I don't think we care about matches, just tasks, so I wouldn't bother to log them.

File: scheduler/src/cook/scheduler/scheduler.clj [hunk 2]
```diff
@@ -1113,219 +1124,220 @@
...
-            (doseq [[compute-cluster offer-subset] (group-by :compute-cluster offers)]
-              (try
-                (cc/restore-offers compute-cluster pool-name offer-subset)
-                (catch Throwable t
-                  (log-structured/error "Error restoring offers for compute cluster"
-                                        {:pool pool-name :compute-cluster compute-cluster}
-                                        t)))))
-          ; if an error happened, it doesn't mean we need to penalize Fenzo
-          true)))))
+    (tracing/with-span [s {:name "handle-resource-offers" :tags {:pool pool-name}}]
```
[the remark this answers is not in the document]
Comment (scrosby) [reply scheduler/src/cook/scheduler/scheduler.clj:1127]: The caller of this, make-offer-handler around line 1357, should have spans encompassing the entire match loop. In fact, that *is* the toplevel span for the match loop. Also a subspan when we do offer generation around line 1386.

File: scheduler/src/cook/scheduler/scheduler.clj [hunk 3]
```diff
@@ -805,64 +808,67 @@
...
-          (ratelimit/spend! compute-cluster-launch-rate-limiter ratelimit/compute-cluster-launch-rate-limiter-key 1))
-        (locking fenzo
-          (-> fenzo
-              (.getTaskAssigner)
-              (.call task-request hostname)))))
-    (catch Throwable t
-      (log-structured/error "Error launching tasks for compute cluster"
-                            {:pool pool-name :compute-cluster (cc/compute-cluster-name compute-cluster)}
-                            t))))
+    (tracing/with-span [s {:name "launch-matches"
```
[the remark this answers is not in the document]
Comment (samincheva) [reply scheduler/src/cook/scheduler/scheduler.clj:811]: They're technically different (launch-tasks vs launch-matches), but I'm not a fan of the ambiguity, either. Will change to be prefixed.

File: scheduler/src/cook/scheduler/scheduler.clj [hunk 4]
```diff
@@ -1053,55 +1063,56 @@
 
 (defn handle-match-cycle-metrics
   [match-map]
-  (let [{:keys [considerable-jobs head-matched? head-resources matches max-considerable
-                number-considerable-jobs number-matched-jobs number-unmatched-jobs offers offers-scheduled
-                pool-name]} match-map
-        user->number-matched-considerable-jobs (->> matches
-                                                    matches->jobs
+  (tracing/with-span [s {:name "handle-match-cycle-metrics"}]
```
[the remark this answers is not in the document]
Comment (samincheva) [reply scheduler/src/cook/scheduler/scheduler.clj:1066]: I had it for completeness, but agree it's not necessary. Will remove.

**Pull request merged**